proposals/state-machine-restriction-updates.md

```diff
@@ -0,0 +1,167 @@
+# State machine restriction updates
+
+## Summary
+[summary]: #summary
+
+- Allow `ref`/`ref struct` locals and `unsafe` blocks in iterators and async methods
+  provided they are used in code segments without any `yield` or `await`.
+- Warn about `yield` inside `lock`.
+- Consolidate spec: disallow pointers to hoisted variables
+  (pointers to captured variables are already disallowed).
+
+## Motivation
+[motivation]: #motivation
+
+It is not necessary to disallow `ref`/`ref struct` locals and `unsafe` blocks in async/iterator methods
+if they are not used across `yield` or `await`, because they do not need to be hoisted.
+
+```cs
+async void M()
+{
+    await ...;
+    ref int x = ...; // error previously, proposed to be allowed
+    x.ToString();
+    await ...;
+    // x.ToString(); // still error
+}
+```
+
+On the other hand, having `yield` inside a `lock` means the caller also holds the lock while iterating which might lead to unexpected behavior.
+This is even more problematic in async iterators where the caller can `await` between iterations, but `await` is not allowed in `lock`.
+See also https://github.com/dotnet/roslyn/issues/72443.
+
+```cs
+lock (this)
+{
+    yield return 1; // warning proposed
+}
+```
+
+## Detailed design
+[design]: #detailed-design
+
+[§13.3.1 Blocks > General][blocks-general]:
+
+> ~~It is a compile-time error for an iterator block to contain an unsafe context ([§23.2][unsafe-contexts]).~~
+> An iterator block always defines a safe context, even when its declaration is nested in an unsafe context.
```

Comment:
nit: use ```` ```diff ```` blocks

Reply:
Maybe it's just me, but for text/spec changes strikethrough renders more nicely (because other text formatting is still applied). Diff blocks are rendered as plain text, which doesn't seem great for this purpose.